### ISSUE TITLE
[Awaiting pending offer] [$125] mWeb - Copilot - User navigated to incorrect page when using device back button in MC page

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#53007]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.66-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** ibellicotest+36@gmail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Open the staging.new.expensify.com website
2. Tap on "Settings" on the bottom of the screen
3. Tap on "Security" and select "Add Copilot"
4. Select any user as copilot
5. Select an access level
6. Tap on "Add Copilot"
7. On Magic Code page, use the device back button to return to previous page
8. Verify you are redirected to copilot confirmation page

## Expected Result:
When using the device back button to return to previous page on magic code page while adding copilot, the user should land on the copilot confirmation page

## Actual Result:
User lands on "Access Level" page instead of copilot confirmation, when using the device back button while in the Magic Code page

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [x] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/a0e22933-33e7-4135-92ac-f3345841c0a0

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021861786354506701472</li>
        <li>Upwork Job ID: 1861786354506701472</li>
        <li>Last Price Increase: 2024-11-27</li>
<li>Automatic offers: </li>
<ul>
<li>dominictb | Reviewer | 105157484</li>
<li>daledah | Contributor | 105157485</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @dominictb</details>

**Comment by twisterdotcom** [comment]
https://github.com/user-attachments/assets/5cb0fbdf-6804-4fa6-8260-27a41c42e12d

**Comment by twisterdotcom** [comment]
Web:

https://github.com/user-attachments/assets/c6a22c6f-34b5-481a-8ef8-31f2f527110b

iOS:

https://github.com/user-attachments/assets/5cb0fbdf-6804-4fa6-8260-27a41c42e12d

**Comment by twisterdotcom** [comment]
Ah, my android device doesn't have a back button, I guess I would need to enable it in the settings. Okay, browser made it much easier. I agree then, this is a bug.

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

When using the device back button while in the Magic Code page, user lands on "Access Level" page instead of copilot confirmation page.

### What is the root cause of that problem?

The RC of the problem is that the copilot confirmation page (`ConfirmDelegatePage`) contains a modal component `DelegateMagicCodeModal` which becomes visible when we tap the `Add copilot` button while the page under the visible modal is the copilot confirmation page (`ConfirmDelegatePage`).

Our issue occurs because the `DelegateMagicCodeModal` does not currently include logic that handles hardware / browser back button in order to return to the copilot confirmation page when going back from the modal.

> [!caution]
> This issue is present on **all platforms where hardware / browser back button is available** - this includes Web / Android: mWeb Chrome / iOS: mWeb Safari.

<details>
<summary>Here's the proof</summary>

| Web: Chrome | iOS: mWeb Safari | Android: mWeb Chrome |
| ------------- | ------------- | ------------- |
| <video src="https://github.com/user-attachments/assets/c5f0130e-5e4c-424b-83a4-f2e49e45c143"> | <video src="https://github.com/user-attachments/assets/726d07a5-5d8b-4f50-8943-017bdfb9e7f6"> | <video src="https://github.com/user-attachments/assets/1fa1ad40-33d4-4a69-bce8-198ca9e031c2"> |

</details>

### What changes do you think we should make in order to solve the problem?

> [!note]
> The reason I decided to post a proposal is because the other proposal would only solve the issue on **mWeb Chrome** which I don't consider a complete solution.

In the `ValidateCodeActionModal` component, pass `shouldHandleNavigationBack` (true) **without any condition**, which would ensure we have the same behaviour of the hardware / browser back button on **all platforms** where this issue is present, as per the **Expected result**.

<details>
<summary>Result videos</summary>

| Web: Chrome | iOS: mWeb Safari | Android: mWeb Chrome |
| ------------- | ------------- | ------------- |
| <video src="https://github.com/user-attachments/assets/088de8cb-2676-4224-af55-3fc52356124e"> | <video src="https://github.com/user-attachments/assets/4aa6655d-41ca-47ce-bd2d-bfe65ef55685"> | <video src="https://github.com/user-attachments/assets/a6f3fe94-3183-4ae9-b29b-2b474e599eee"> |

</details>

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

When using the device back button while in the Magic Code page, user lands on "Access Level" page instead of copilot confirmation page.

### What is the root cause of that problem?

The RC of the problem is that the copilot confirmation page (`ConfirmDelegatePage`) contains a modal component `DelegateMagicCodeModal` which becomes visible when we tap the `Add copilot` button while the page under the visible modal is the copilot confirmation page (`ConfirmDelegatePage`).

Our issue occurs because the `DelegateMagicCodeModal` does not currently include logic that handles hardware / browser back button in order to return to the copilot confirmation page when going back from the modal.

> [!caution]
> This issue is present on **all platforms where hardware / browser back button is available** - this includes Web / Android: mWeb Chrome / iOS: mWeb Safari.

<details>
<summary>Here's the proof</summary>

| Web: Chrome | iOS: mWeb Safari | Android: mWeb Chrome |
| ------------- | ------------- | ------------- |
| <video src="https://github.com/user-attachments/assets/c5f0130e-5e4c-424b-83a4-f2e49e45c143"> | <video src="https://github.com/user-attachments/assets/726d07a5-5d8b-4f50-8943-017bdfb9e7f6"> | <video src="https://github.com/user-attachments/assets/1fa1ad40-33d4-4a69-bce8-198ca9e031c2"> |

</details>

### What changes do you think we should make in order to solve the problem?

> [!note]
> 1. The reason I decided to post a proposal is because the other proposal would only solve the issue on **mWeb Chrome** which I don't consider a complete solution.
> 2. Another error of the other proposal is that `DelegateMagicCodeModal` does not take the `shouldHandleNavigationBack` prop dirrectly.

In the `ValidateCodeActionModal` component, pass `shouldHandleNavigationBack` (true) **without any condition**, which would ensure we have the same behaviour of the hardware / browser back button on **all platforms** where this issue is present, as per the **Expected result**.

<details>
<summary>Result videos</summary>

| Web: Chrome | iOS: mWeb Safari | Android: mWeb Chrome |
| ------------- | ------------- | ------------- |
| <video src="https://github.com/user-attachments/assets/088de8cb-2676-4224-af55-3fc52356124e"> | <video src="https://github.com/user-attachments/assets/4aa6655d-41ca-47ce-bd2d-bfe65ef55685"> | <video src="https://github.com/user-attachments/assets/a6f3fe94-3183-4ae9-b29b-2b474e599eee"> |

</details>

**Comment by MarioExpensify** [comment]
Nice!! Thank you @dominictb, let's move forward with @daledah proposal.

**Comment by MarioExpensify** [comment]
> Use `Browser.isMobileChrome()` if we only want to fix it in mobile chrome.

@daledah I don't think that is the intention, we should have this fixed in native with back buttons as well, @twisterdotcom please correct me if I got this wrong.

**Comment by twisterdotcom** [comment]
Sounds good.

**Comment by twisterdotcom** [comment]
Payment Summary:

- @dominictb paid $125 [here](https://www.upwork.com/nx/wm/offer/105157484) (Reviewer)
- @daledah paid $125 [here](https://www.upwork.com/nx/wm/offer/105157485) (Contributor)

**Comment by twisterdotcom** [comment]
Waiting on @dominictb to accept the Upwork offer now.

**Comment by MarioExpensify** [comment]
I think this can be closed now, but I'll wait @twisterdotcom to be back
